### PR TITLE
fix: handle calldata tags in custom approvals

### PR DESCRIPTION
**Pull request by horsefacts**
## **Description**

The custom token approval utility function in [`ui/pages/confirmations/confirm-approve/confirm-approve.util.js`](https://github.com/MetaMask/metamask-extension/compare/develop...horsefacts:horsefacts/approval-calldata-tags?expand=1#diff-2df42ad862369436640f1683e9c5408b24f418028dad44840da1609587b6938c) parses ERC20 `approve` calldata and reconstructs it with an optional custom value.

This utility validates that the length of the parsed `value` amount is exactly 64 hex digits, assuming all calldata bytes after the `spender` are the `value` and the value is a `uint256`. However, this validation is too strict: it's possible for the calldata to include extra bytes following the value, which the contract will ignore in the call to `approve`. These are valid transactions. 

Including a "calldata tag" for onchain attribution or analytics is an increasingly popular pattern. For example, [Seaport](https://opensea.notion.site/Seaport-Order-Attributions-ec2d69bf455041a5baa490941aad307f) uses calldata tags for order attribution, [Viem](https://viem.sh/docs/contract/writeContract#datasuffix) provides a `dataSuffix` option, and Farcaster clients append an [attribution suffix](https://warpcast.notion.site/Frame-Transactions-Public-9d9f9f4f527249519a41bd8d16165f73#ef8e2e7eac9a42808087ee30c8ff15f5) to transactions from Farcaster frames.

This utility should instead parse the first 64 bytes of the `value` and preserve any extra calldata.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/23527?quickstart=1)

## **Screenshots**

Error screenshot:

error image

## **Related issues**

Previous issue in seaport-js: https://github.com/ProjectOpenSea/seaport-js/issues/215

## **Manual testing steps**

1. Create an ERC20 `approve` with a calldata suffix. For example, using Viem:

```ts
await walletClient.writeContract({
  address: '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48',
  abi: erc20Abi,
  functionName: 'approve',
  args: ['0x0000000000000000000000000000000000000001', 69420],
  dataSuffix: '0xdeadbeef'
});
```

2. Execute the transaction in Metamask.

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by horsefacts** [comment]
I have read the CLA Document and I hereby sign the CLA

File: ui/pages/confirmations/confirm-approve/confirm-approve.util.js [hunk 1]
```diff
@@ -21,16 +21,14 @@ export function getCustomTxParamsData(
   if (spender.startsWith('0x')) {
     spender = spender.substring(2);
   }
-  const [signature, tokenValue] = data.split(spender);
-
-  if (!signature || !tokenValue) {
+  const [signature, rest] = data.split(spender);
+  if (!signature || !rest) {
     throw new Error('Invalid data');
-  } else if (tokenValue.length !== 64) {
```
Comment (adonesky1) [inline ui/pages/confirmations/confirm-approve/confirm-approve.util.js:28]: Perhaps we ought to continue to throw if `rest` value is less than 64 characters?

Comment (adonesky1) [reply ui/pages/confirmations/confirm-approve/confirm-approve.util.js:35]: Could we perhaps add the logic here just in case that upstream code changes?

Comment (horsefacts) [reply ui/pages/confirmations/confirm-approve/confirm-approve.util.js:35]: Sure, just added this check back!

File: ui/pages/confirmations/confirm-approve/confirm-approve.util.js [hunk 2]
```diff
@@ -28,16 +28,16 @@ export function getCustomTxParamsData(
...
-    throw new Error(
-      'Invalid token value; should be exactly 64 hex digits long (u256)',
-    );
+  } else if (rest.length < 64) {
+    throw new Error('Invalid value; must be at least 64 hex digits long');
```
Comment (adonesky1) [inline ui/pages/confirmations/confirm-approve/confirm-approve.util.js:35]: [nit]
```suggestion
    throw new Error('Invalid calldata value; must be at least 64 hex digits long');
```

**Review by adonesky1 — DISMISSED** [review]

**Review by jpuri — DISMISSED** [review]
lgtm

**Review by adonesky1 — APPROVED** [review]

**Review by jpuri — APPROVED** [review]

**Pull request merged**